explorer/interpreter/stack_space.h

```diff
@@ -0,0 +1,57 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_EXPLORER_INTERPRETER_STACK_SPACE_H_
+#define CARBON_EXPLORER_INTERPRETER_STACK_SPACE_H_
+
+#include <functional>
+#include <optional>
+
+namespace Carbon {
+
+namespace Internal {
+
+// Returns true if a new thread should be started for more stack space.
+auto IsStackSpaceNearlyExhausted() -> bool;
+
+// Starts a thread to run the function.
+auto RunWithStackSpaceHelper(std::function<void()> fn) -> void;
+
+}  // namespace Internal
+
+// Starts a new thread with initialized stack space. This should be called
+// before RunWithStackSpace so that execution is occurring with known stack
+// space.
+//
+// Usage:
+//   return InitStackSpace<ReturnType>([&]() -> ReturnType {
+//         <function body>
+//       });
+template <typename ReturnType>
+auto InitStackSpace(std::function<ReturnType()> fn) -> ReturnType {
+  std::optional<ReturnType> result;
+  Internal::RunWithStackSpaceHelper([&] { result = fn(); });
+  return std::move(*result);
+}
+
+// Runs a function. May start a thread if more stack space is desirable.
```

Comment:
I thought our goal here was to detect when the Carbon program is consuming too many resources, so that we don't waste too much time when the fuzzer discovers how to create an infinite loop in the Carbon code. This seems like it's doing the opposite, enabling us to spend more resources in those cases. Have I misunderstood the intent?